cpp/src/bitmask/bit_container.cuh

```diff
@@ -0,0 +1,223 @@
+  }
+
+  /**
+   * Set a bit in a thread-safe manner
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void SetBit(T record_idx) {
+    int rec = bit_container::which_word(record_idx);
+    int bit = bit_container::which_bit(record_idx);
+
+    atomicOr( &valid_[rec], (1U << bit));
+  }
+
+
+  /**
+   * Clear a bit in a thread-safe manner
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void ClearBit(T record_idx) {
+    int rec = bit_container::which_word(record_idx);
+    int bit = bit_container::which_bit(record_idx);
+
+    atomicAnd( &valid_[rec], ~(1U << bit));
+  }
+
+  /**
+   * Get the number of words in this bit container
+   *
+   * @return the number of words
+   */
+  __device__ int NumWords() const {
+    return bit_container::num_words(bitlength_);
+  }
+
+  /**
+   * Getter for valid
+   *
+   * @return pointer to valid bit container
+   */
+  __host__ __device__ bit_container_t *GetValid() const {
+    return valid_;
+  }
+
+  /**
+   * Get length
+   *
+   * @return length of bit container in bits
+   */
+  __host__ __device__ int Length() const {
+    return bitlength_;
+  }
+
+
+  /**
+   * Allocate device space for the valid bitmap.
+   *
+   * @param[in]  number_of_records     number of records
+   * @param[in]  fill_value            optional, should the memory be initialized to all 0 or 1s. All other values indicate un-initialized
+   * @return pointer to allocated space on the device
+   */
+  template <typename T>
+  __host__
+  static bit_container_t * CreateBitContainer(T number_of_records, int fill_value = -1) {
+    bit_container_t *container_d = nullptr;
+    int num_bit_containers = bit_container::num_words(number_of_records);
+
+    RMM_ALLOC((void**)&container_d, sizeof(bit_container_t) * num_bit_containers, 0);
+    if (container_d == nullptr)
+      return container_d;
+
+    if (fill_value == 0) {
+      cudaMemset(container_d, 0, sizeof(bit_container_t) * num_bit_containers);
+    } else if (fill_value == 1) {
+      cudaMemset(container_d, 0xff, sizeof(bit_container_t) * num_bit_containers);
+
+      //
+      //  Need to worry about the case where we set bits past the end
+      //
+      int bits_used_in_last_word = number_of_records - (num_bit_containers - 1) * bit_container::BITS_PER_WORD;
+      bit_container_t temp = 1U << bits_used_in_last_word;
+      temp--;
+
+      PutWord(temp, &container_d[num_bit_containers - 1]);
+    }
+
+    return container_d;
+  }
+
+  /**
+   *  Copy data between host and device
+   *
+   *  @param[out] dst      - the address of the destination
+   *  @param[in] src       - the address of the source
+   *  @param[in] num_bits  - the number of bits in the bit container
+   *  @param[in] kind      - the direction of the copy
+   */
+  inline __host__
+  static void CopyBitContainer(bit_container_t *dst, const bit_container_t *src, size_t num_bits, enum cudaMemcpyKind kind) {
```

Comment:
I don't think this function is necessary. It doesn't add anything over having people just call `cudaMemcpy` directly.

Reply:
This is more a design philosophy question.  I agree that the user can just as easily call cudaMemcpy directly.  The question is do we want the user to understand the internal workings of the class, or do we want to provide them with a layer that protects them from changes we might make in a future version (perhaps if the memory model changes in the future).

Now, it's not clear to me that this particular call really protects the users from those changes or not.